src/pip/_internal/req/req_install.py

```diff
@@ -67,7 +92,8 @@ def __init__(self, req, comes_from, source_dir=None, editable=False,
         if link is not None:
             self.link = self.original_link = link
         else:
-            self.link = self.original_link = req and req.url and Link(req.url)
+            # https://github.com/python/mypy/issues/5540
+            self.link = self.original_link = req and req.url and Link(req.url)  # type: ignore  # noqa: E501
```

Comment:
I would suggest making this a conditional assignment, replacing the else block with:

```py
elif req and req.url:
    # PEP 508 URL requirement
    self.link = Link(req.url)
else:
    self.link = None
```